src/Tests/Service/RegistrationServiceTest.php

```diff
@@ -0,0 +1,197 @@
+        $response->clientData = 'client-data';
+
+        $u2f = m::mock('u2flib_server\U2F');
+        $u2f->shouldReceive('doRegister')
+            ->once()
+            ->with(m::anyOf($yubicoRequest), m::anyOf($response))
+            ->andThrow(new Error('error', $errorCode));
+
+        $service = new RegistrationService($u2f);
+
+        $this->assertEquals($expectedResult, $service->verifyRegistration($request, $response));
+    }
+
+    public function expectedVerificationErrors()
+    {
+        // Autoload the U2F class to make sure the error constants are loaded which are also defined in the file.
+        class_exists('u2flib_server\U2F');
+
+        return [
+            'responseChallengeDidNotMatchRequestChallenge' => [
+                \u2flib_server\ERR_UNMATCHED_CHALLENGE,
+                RegistrationVerificationResult::responseChallengeDidNotMatchRequestChallenge()
+            ],
+            'responseWasNotSignedByDevice' => [
+                \u2flib_server\ERR_ATTESTATION_SIGNATURE,
+                RegistrationVerificationResult::responseWasNotSignedByDevice()
+            ],
+            'deviceCannotBeTrusted' => [
+                \u2flib_server\ERR_ATTESTATION_VERIFICATION,
+                RegistrationVerificationResult::deviceCannotBeTrusted()
+            ],
+            'publicKeyDecodingFailed' => [
+                \u2flib_server\ERR_PUBKEY_DECODE,
+                RegistrationVerificationResult::publicKeyDecodingFailed()
+            ],
+        ];
+    }
+
+    /**
+     * @test
+     * @group registration
+     * @dataProvider unexpectedVerificationErrors
+     *
+     * @param int $errorCode
+     */
+    public function it_throws_unexpected_u2f_registration_verification_errors($errorCode)
+    {
+        $yubicoRequest = new \u2flib_server\RegisterRequest('challenge', self::APP_ID);
+
+        $request = new \Surfnet\StepupU2fBundle\Dto\RegisterRequest();
+        $request->version   = 'U2F_V2';
+        $request->challenge = 'challenge';
+        $request->appId     = self::APP_ID;
+
+        $response = new \Surfnet\StepupU2fBundle\Dto\RegisterResponse();
+        $response->registrationData = 'registration-data';
+        $response->clientData = 'client-data';
+
+        $u2f = m::mock('u2flib_server\U2F');
+        $u2f->shouldReceive('doRegister')
+            ->once()
+            ->with(m::anyOf($yubicoRequest), m::anyOf($response))
+            ->andThrow(new Error('error', $errorCode));
+
+        $service = new RegistrationService($u2f);
+
+        $this->setExpectedExceptionRegExp('Surfnet\StepupU2fBundle\Exception\LogicException');
+        $service->verifyRegistration($request, $response);
+    }
+
+    public function unexpectedVerificationErrors()
+    {
+        // Autoload the U2F class to make sure the error constants are loaded which are also defined in the file.
+        class_exists('u2flib_server\U2F');
+
+        return [
+            [\u2flib_server\ERR_AUTHENTICATION_FAILURE],
```

Comment:
This error code is used for exceptions during authentication only. If it occurs, there are some serious `LogicException`s going on in a programmer.